hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -484,30 +488,64 @@ def greedy_shrink(self):
         it twice will have exactly the same effect as calling it once.
         """
 
-        self.fixate_shrink_passes(
-            [
-                block_program("X" * 5),
-                block_program("X" * 4),
-                block_program("X" * 3),
-                block_program("X" * 2),
-                block_program("X" * 1),
-                "pass_to_descendant",
-                "adaptive_example_deletion",
-                "alphabet_minimize",
-                "zero_examples",
-                "reorder_examples",
-                "minimize_floats",
-                "minimize_duplicated_blocks",
-                block_program("-XX"),
-                "minimize_individual_blocks",
-                block_program("--X"),
-            ]
-        )
+        passes = [
+            block_program("X" * 5),
+            block_program("X" * 4),
+            block_program("X" * 3),
+            block_program("X" * 2),
+            block_program("X" * 1),
+            "pass_to_descendant",
+            "adaptive_example_deletion",
+            "alphabet_minimize",
+            "zero_examples",
+            "reorder_examples",
+            "minimize_floats",
+            "minimize_duplicated_blocks",
+            block_program("-XX"),
+            "minimize_individual_blocks",
+            block_program("--X"),
+        ]
+
+        while True:
+            selected = self.select_shrink_passes(passes)
+            if not selected:
+                break
+            self.debug("Selected passes %s" % (", ".join(sp.name for sp in selected)))
+            self.fixate_shrink_passes(selected)
 
     @derived_value
     def shrink_pass_choice_trees(self):
         return defaultdict(ChoiceTree)
 
+    def select_shrink_passes(self, passes):
+        """Select a subset of ``passes`` that it currently makes sense to run.
+
+        Will return an empty list if and only if no pass in the list can
+        currently make progress.
+        """
+
+        passes = list(map(self.shrink_pass, passes))
+
+        any_ran = True
+        max_failures = 3
+        while any_ran:
+            any_ran = False
+            selected = []
+            for sp in passes:
+                initial_calls = self.calls
+                while self.calls < initial_calls + max_failures:
+                    prev = self.shrink_target
+                    if not sp.step(random=True):
+                        break
+                    any_ran = True
+                    if prev is not self.shrink_target:
+                        selected.append(sp)
+                        break
+            if selected:
+                return selected
+            else:
+                max_failures *= 2
+
```

Comment:
Is this supposed to have an explicit return for the failure case?

Right now it looks like it just falls off and implicitly returns `None`, which happens to work but is not very clear (and contradicts the docstring).